src/Checkpointer.jl

```diff
@@ -78,18 +96,188 @@ end
+                cs.comms_ctx,
+                Int((t - t0) / 1e3),
+                output_dir = cs.dirs.checkpoints,
+            )
+        end
+    end
+
+    # Checkpoint the Coupler fields
+    output_dir = cs.dirs.checkpoints
+    comms_ctx = cs.comms_ctx
+    time = Int((t - t0) / 1e3)
+    day = floor(Int, time / (60 * 60 * 24))
+    sec = floor(Int, time % (60 * 60 * 24))
+    pid = ClimaComms.mypid(comms_ctx)
+    @info "Saving coupler fields to JLD2 on day $day second $sec"
+    output_file = joinpath(output_dir, "checkpoint_coupler_fields_$(pid)_$time.jld2")
+    JLD2.jldsave(output_file, coupler_fields = CC.Adapt.adapt(Array, cs.fields))
+end
+
+"""
+    auto_restart!(cs::CoupledSimulation, checkpoint_dir)::Bool
+
+Try finding restart files for `cs` in `checkpoint_dir` and use them to reset the
+content of `cs`.
+
+Return a true if the simulation was restarted.
+"""
+function auto_restart!(cs, checkpoint_dir)
+    comms_ctx = cs.comms_ctx
+    for sim in cs.model_sims
+        if Checkpointer.get_model_prog_state(sim) !== nothing
+            auto_restart_model_state!(sim, comms_ctx, checkpoint_dir)
+        end
+        if Checkpointer.get_model_cache(sim) !== nothing
+            auto_restart_model_cache!(sim, comms_ctx, checkpoint_dir)
+        end
+    end
+    auto_restart_coupler_fields!(cs, checkpoint_dir)
+end
+
+"""
+    restart!(cs::CoupledSimulation, checkpoint_dir, checkpoint_t)
+
+Overwrite the content of `cs` with checkpoints in `checkpoint_dir` at time `checkpoint_t`.
+
+Return a true if the simulation was restarted.
+"""
+function restart!(cs, checkpoint_dir, checkpoint_t)
+    pid = ClimaComms.mypid(cs.comms_ctx)
+    for sim in cs.model_sims
+        if Checkpointer.get_model_prog_state(sim) !== nothing
+            input_file_state =
+                output_file = joinpath(checkpoint_dir, "checkpoint_$(Interfacer.name(sim))_$(checkpoint_t).hdf5")
+            restart_model_state!(sim, input_file_state, cs.comms_ctx)
+            input_file_cache =
+                joinpath(checkpoint_dir, "checkpoint_cache_$(pid)_$(Interfacer.name(sim))_$(checkpoint_t).jld2")
+            restart_model_cache!(sim, input_file_cache)
         end
     end
+    input_file_coupler_fields = joinpath(checkpoint_dir, "checkpoint_coupler_fields_$(pid)_$(checkpoint_t).jld2")
+    restart_coupler_fields!(cs, input_file_coupler_fields)
+    return true
+end
+
+"""
+    auto_restart_coupler_fields!(cs, checkpoint_dir)
+
+Find checkpoints in `checkpoint_dir` for the coupler fields and write them in `cs`.
+"""
+function auto_restart_coupler_fields!(cs, checkpoint_dir)
+    isdir(checkpoint_dir) || return false
+    pid = ClimaComms.mypid(cs.comms_ctx)
+    restart_file_rx = Regex("checkpoint_coupler_fields_$(pid)_\\d+.jld2")
+    restarts = filter(f -> !isnothing(match(restart_file_rx, f)), readdir(checkpoint_dir))
+    isempty(restarts) && return false
+    input_file = joinpath(checkpoint_dir, last(sort_by_creation_time(restarts)))
```

Comment:
Could you explain the behavior more in the docstring please? E.g. if there are multiple files in the checkpoint directory matching the expected format, the most recent file will be used to restart

(same comment applies to `auto_restart_model_cache!`, `auto_restart_model_state!`)

Reply:
I removed this function because at the end of the day I could accomplish the same behavior with the `t_start_from_checkpoint` function